src/main/java/net/logstash/logback/appender/AbstractLogstashTcpSocketAppender.java

```diff
@@ -531,56 +524,68 @@ public void onEvent(LogEvent<Event> logEvent, long sequence, boolean endOfBatch)
+                 *
+                 * If socket == null here, it means that a write timed out,
+                 * and the socket was closed by the WriteTimeoutRunnable.
+                 *
+                 * Therefore, attempt reconnection.
+                 */
                 Future<?> readerFuture = this.readerFuture;  // volatile read
                 if (readerFuture.isDone() || socket == null) {
-                    /*
-                     * If readerFuture.isDone(), then the destination has shut down its output (our input),
-                     * and the destination is probably no longer listening to its input (our output).
-                     * This will be the case for Amazon's Elastic Load Balancers (ELB)
-                     * when an instance behind the ELB becomes unhealthy while we're connected to it.
-                     *
-                     * If socket == null here, it means that a write timed out,
-                     * and the socket was closed by the WriteTimeoutRunnable.
-                     *
-                     * Therefore, attempt reconnection.
-                     */
                     addInfo(peerId + "destination terminated the connection. Reconnecting.");
                     reopenSocket();
-                    try {
-                        readerFuture.get();
-                        sendFailureException = NOT_CONNECTED_EXCEPTION;
-                    } catch (Exception e) {
-                        sendFailureException = e;
-                    }
+                    
+// TODO readerFuture refers to the "old" future and a new one has been started by reopenSocket...
+// Why wait for the old to complete? Is it to reuse its thread and avoid creating/wasting a new one?
+// In this case better to readerFuture.get() before reopenSocket()...
+//
+//                    try {
+//                        readerFuture.get();
+//                        sendFailureException = NOT_CONNECTED_EXCEPTION;
+//                    } catch (Exception e) {
+//                        sendFailureException = e;
+//                    }
                     continue;
                 }
+                
+                /*
+                 * Write the event in the output stream.
+                 * Drop event if encoder throws an exception.
+                 * Reconnect if an exception is thrown by the connection stream itself.
+                 */
                 try {
                     writeEvent(socket, outputStream, logEvent, endOfBatch);
                     return;
+                    
+                } catch (EncoderException e) {
+                    /*
+                     * Encoding threw an exception. Warn and drop event before it becomes a "poison".
+                     */
+                    addWarn(peerId + "Encoder failed to encode event: " + e.getMessage() + ". Dropping event.", e);
+                    fireEventSendFailure(logEvent.event, e);
+                    break;
```

Comment:
I think this should log and fire the EncoderException's cause, rather than the EncoderException itself, since EncoderException is private, and is just used internally to detect where the exception occurred.